compiler/src/iree/compiler/Codegen/Common/EmulateNarrowType.cpp

```diff
@@ -36,17 +38,49 @@ struct ConvertHalInterfaceBindingSubspan final
   LogicalResult
   matchAndRewrite(IREE::HAL::InterfaceBindingSubspanOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
-    Type newResultTy = getTypeConverter()->convertType(op.getType());
-    if (!newResultTy)
+    auto currentType = op.getType().dyn_cast<MemRefType>();
+    if (!currentType) {
+      return rewriter.notifyMatchFailure(op->getLoc(),
+                                         "unhandled non-memref types");
+    }
+    auto newResultType =
+        getTypeConverter()->convertType(currentType).dyn_cast<MemRefType>();
+    if (!newResultType) {
       return rewriter.notifyMatchFailure(
           op->getLoc(),
           llvm::formatv("failed to legalize memref type: {0}", op.getType()));
+    }
+    Location loc = op.getLoc();
+    OpFoldResult zero = rewriter.getIndexAttr(0);
+    SmallVector<OpFoldResult> indices(currentType.getRank(), zero);
+
+    // Get linearized type.
+    int srcBits = currentType.getElementType().getIntOrFloatBitWidth();
+    int dstBits = newResultType.getElementType().getIntOrFloatBitWidth();
+    OpFoldResult elementOffset;
+    Value byteOffset = adaptor.getByteOffset();
+    if (byteOffset && !matchPattern(byteOffset, m_Zero())) {
+      elementOffset = convertByteOffsetToElementOffset(
+          rewriter, loc, byteOffset, currentType.getElementType());
+    } else {
+      elementOffset = rewriter.getIndexAttr(0);
+    }
```

Comment:
[optional] I'd prefer having `zero` be elementOffset's default value. The logic makes sense toe me and it saves us 2 lines.

```suggestion
    OpFoldResult elementOffset = rewriter.getIndexAttr(0);
    Value byteOffset = adaptor.getByteOffset();
    if (byteOffset && !matchPattern(byteOffset, m_Zero())) {
      elementOffset = convertByteOffsetToElementOffset(
          rewriter, loc, byteOffset, currentType.getElementType());
    }
```